server/src/main/java/org/opensearch/index/search/comparators/UnsignedLongComparator.java

```diff
@@ -0,0 +1,107 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.index.search.comparators;
+
+import org.apache.lucene.index.LeafReaderContext;
+import org.apache.lucene.sandbox.document.BigIntegerPoint;
+import org.apache.lucene.search.LeafFieldComparator;
+import org.apache.lucene.search.comparators.NumericComparator;
+import org.opensearch.common.Numbers;
+
+import java.io.IOException;
+import java.math.BigInteger;
+
+/** The comparator for unsigned long numeric type */
+public class UnsignedLongComparator extends NumericComparator<BigInteger> {
+    private final BigInteger[] values;
+    protected BigInteger topValue;
+    protected BigInteger bottom;
+
+    public UnsignedLongComparator(int numHits, String field, BigInteger missingValue, boolean reverse, boolean enableSkipping) {
+        super(field, missingValue != null ? missingValue : Numbers.MIN_UNSIGNED_LONG_VALUE, reverse, enableSkipping, BigIntegerPoint.BYTES);
+        values = new BigInteger[numHits];
+    }
+
+    @Override
+    public int compare(int slot1, int slot2) {
+        return values[slot1].compareTo(values[slot2]);
+    }
+
+    @Override
+    public void setTopValue(BigInteger value) {
+        super.setTopValue(value);
+        topValue = value;
+    }
+
+    @Override
+    public BigInteger value(int slot) {
+        return values[slot];
+    }
+
+    @Override
+    public LeafFieldComparator getLeafComparator(LeafReaderContext context) throws IOException {
+        return new UnsignedLongLeafComparator(context);
+    }
+
+    /** Leaf comparator for {@link UnsignedLongComparator} that provides skipping functionality */
+    public class UnsignedLongLeafComparator extends NumericLeafComparator {
+
+        public UnsignedLongLeafComparator(LeafReaderContext context) throws IOException {
+            super(context);
+        }
+
+        private BigInteger getValueForDoc(int doc) throws IOException {
+            if (docValues.advanceExact(doc)) {
+                return Numbers.toUnsignedBigInteger(docValues.longValue());
```

Comment:
Curious, lucene only supports value reading for long, what happens to values higher that Long.Max, I mean, lucene itself will be having precision rounding here right ?

Reply:
So for long we should not have precision issues but only sign changes (negative / positive)